### ISSUE TITLE
userns-remap and file ownership inside the container

### ISSUE THREAD

**Issue by akamalov** [docker/docker#31332]
**Description**

Running Docker with ```--userns-remap=default```.  ```/etc/subgid``` and ```/etc/subuid``` are set to ```dockremap:100000:65536```. When container gets run in HOST mode, files inside the container are owned by ```100000:100000```  

**Steps to reproduce the issue:**

```
[root@node237 ~]# docker run -it --net=host --userns=host nginx bash
root@node237:/# ls -al
total 20
drwxr-xr-x  21 100000 100000 4096 Feb 24 13:31 .
drwxr-xr-x  21 100000 100000 4096 Feb 24 13:31 ..
-rwxr-xr-x   1 100000 100000    0 Feb 24 13:31 .dockerenv
drwxr-xr-x   2 100000 100000 4096 Dec 13 19:09 bin
drwxr-xr-x   2 100000 100000    6 Sep 12 04:09 boot
drwxr-xr-x   5 root   root    380 Feb 24 13:31 dev
drwxr-xr-x  46 100000 100000 4096 Feb 24 13:31 etc
drwxr-xr-x   2 100000 100000    6 Sep 12 04:09 home
drwxr-xr-x   9 100000 100000  105 Dec 27 17:56 lib
drwxr-xr-x   2 100000 100000   33 Dec 13 19:08 lib64
drwxr-xr-x   2 100000 100000    6 Dec 13 19:07 media
drwxr-xr-x   2 100000 100000    6 Dec 13 19:07 mnt
drwxr-xr-x   2 100000 100000    6 Dec 13 19:07 opt
dr-xr-xr-x 340 root   root      0 Feb 24 13:31 proc
drwx------   2 100000 100000   35 Dec 13 19:07 root
drwxr-xr-x   3 100000 100000   28 Dec 13 19:07 run
drwxr-xr-x   2 100000 100000 4096 Dec 13 19:09 sbin
drwxr-xr-x   2 100000 100000    6 Dec 13 19:07 srv
dr-xr-xr-x  13 root   root      0 Feb 23 17:12 sys
drwxrwxrwt   2 100000 100000    6 Dec 27 17:56 tmp
drwxr-xr-x  10 100000 100000   97 Dec 27 17:56 usr
drwxr-xr-x  11 100000 100000  128 Dec 27 17:56 var
root@node237:/#
```
Now, let's deploy nginx on SDN subsystem. We're using Nuage Networks

```
[root@node237 ~]# docker run -it --net="Test-lab" nginx bash
```
List files

```
root@c529af1c61bd:/# ls -al
total 20
drwxr-xr-x  21 root   root    4096 Feb 24 13:35 .
drwxr-xr-x  21 root   root    4096 Feb 24 13:35 ..
-rwxr-xr-x   1 root   root       0 Feb 24 13:35 .dockerenv
drwxr-xr-x   2 root   root    4096 Dec 13 19:09 bin
drwxr-xr-x   2 root   root       6 Sep 12 04:09 boot
drwxr-xr-x   5 root   root     380 Feb 24 13:35 dev
drwxr-xr-x  46 root   root    4096 Feb 24 13:35 etc
drwxr-xr-x   2 root   root       6 Sep 12 04:09 home
drwxr-xr-x   9 root   root     105 Dec 27 17:56 lib
drwxr-xr-x   2 root   root      33 Dec 13 19:08 lib64
drwxr-xr-x   2 root   root       6 Dec 13 19:07 media
drwxr-xr-x   2 root   root       6 Dec 13 19:07 mnt
drwxr-xr-x   2 root   root       6 Dec 13 19:07 opt
dr-xr-xr-x 344 nobody nogroup    0 Feb 24 13:35 proc
drwx------   2 root   root      35 Dec 13 19:07 root
drwxr-xr-x   3 root   root      28 Dec 13 19:07 run
drwxr-xr-x   2 root   root    4096 Dec 13 19:09 sbin
drwxr-xr-x   2 root   root       6 Dec 13 19:07 srv
dr-xr-xr-x  13 nobody nogroup    0 Feb 24 13:35 sys
drwxrwxrwt   2 root   root       6 Dec 27 17:56 tmp
drwxr-xr-x  10 root   root      97 Dec 27 17:56 usr
drwxr-xr-x  11 root   root     128 Dec 27 17:56 var
root@c529af1c61bd:/#
```
File owernshipt is the way it should be.

The example shown above is done using nginx, but we're trying to deploy AVI
Networks service engines which while running in host-mode with ```100000:100000``` file ownership,
they're failing to talk to controller if ```userns-remap``` is enabled.

Any feedback or information is welcome.

**Output of `docker version`:**
```
[root@node237 ~]# docker version
Client:
 Version:      1.12.6
 API version:  1.24
 Go version:   go1.6.4
 Git commit:   78d1802
 Built:        Tue Jan 10 20:20:01 2017
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.6
 API version:  1.24
 Go version:   go1.6.4
 Git commit:   78d1802
 Built:        Tue Jan 10 20:20:01 2017
 OS/Arch:      linux/amd64
[root@node237 ~]#
```

**Output of `docker info`:**

```
[root@node237 ~]# docker info
Containers: 8
 Running: 6
 Paused: 0
 Stopped: 2
Images: 196
Server Version: 1.12.6
Storage Driver: devicemapper
 Pool Name: docker-thinpool
 Pool Blocksize: 524.3 kB
 Base Device Size: 10.74 GB
 Backing Filesystem: xfs
 Data file:
 Metadata file:
 Data Space Used: 17.49 GB
 Data Space Total: 51 GB
 Data Space Available: 33.51 GB
 Metadata Space Used: 5.939 MB
 Metadata Space Total: 532.7 MB
 Metadata Space Available: 526.7 MB
 Thin Pool Minimum Free Space: 5.1 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: true
 Deferred Deletion Enabled: true
 Deferred Deleted Device Count: 0
 Library Version: 1.02.107-RHEL7 (2015-10-14)
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: host overlay nuage bridge null
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: seccomp
Kernel Version: 3.10.0-327.4.4.el7.x86_64
Operating System: Red Hat Enterprise Linux Server 7.2 (Maipo)
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 11.58 GiB
Name: node237.localnet.net
ID: 2ET2:UGE4:ZGQP:KAOS:HVHP:IU3T:TFGC:7Z6Y:3ICI:XQLK:DAXF:DTJK
Docker Root Dir: /var/lib/docker/100000.100000
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 73
 Goroutines: 98
 System Time: 2017-02-24T08:41:31.145413876-05:00
 EventsListeners: 1
Username: twistlockreader
Registry: https://index.docker.io/v1/
Cluster Store: consul://192.168.120.156:8500
Cluster Advertise: 192.168.1.237:2376
Insecure Registries:
 node237.localnet.net:5000
 127.0.0.0/8
```

**Comment by cpuguy83** [comment]
I believe this is just a side-effect of how userns support is implemented in Docker.
When images are ingested into docker (via `pull`, `load`, whatever) the content is chowned to match the remapped uid/gids.
When you set `--userns=host` you are disabling the user namespace but the files are still going to be chowned to the remapped users.

The reason it's done this way is because otherwise everything would have to be chowned at runtime which:
1. Would be incredibly slow
2. Would defeat the layered storage

The kernel does not currently have a way to remap uid/gid's on files in a namespace. This is being worked on, see: https://lkml.org/lkml/2017/2/4/138
Once this, or something like this, is merged, we'll be able to dynamically remap file ownership at runtime based on the namespace used in the container.

Closing since this is non-actionable and tracked elsewhere, but feel free to discuss, happy to answer any questions.
Thanks!
